Migrate mixed join to use multiset

PointKernel opened the pull request:
## Description

This PR migrates mixed join to use the new `cuco::static_multiset` data structure and fixes several small issues in the mixed semi join code. This is the last PR to close #12261.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

PointKernel commented:
/ok to test 2f2110c

PointKernel commented:
/ok to test a9d0805

PointKernel commented:
```
['mixed-join-ref.json', 'mixed-join-latest.json']
# mixed_inner_join

## [0] Quadro RTX 8000

|  Nullable  |  NullEquality  |  DataType  |  left_size  |  right_size  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |         Diff |   %Diff |  Status  |
|------------|----------------|------------|-------------|--------------|------------|-------------|------------|-------------|--------------|---------|----------|
|     0      | NULLS_UNEQUAL  |   INT32    |    1000     |     1000     | 146.165 us |       8.80% | 133.259 us |       6.68% |   -12.906 us |  -8.83% |   FAST   |
|     0      | NULLS_UNEQUAL  |   INT32    |   100000    |     1000     | 179.619 us |       7.47% | 167.898 us |       2.12% |   -11.721 us |  -6.53% |   FAST   |
|     0      | NULLS_UNEQUAL  |   INT32    |  10000000   |     1000     |   4.880 ms |       0.47% |   5.122 ms |       0.47% |   241.997 us |   4.96% |   SLOW   |
|     0      | NULLS_UNEQUAL  |   INT32    |   100000    |    100000    | 199.140 us |       2.58% | 191.086 us |       1.77% |    -8.053 us |  -4.04% |   FAST   |
|     0      | NULLS_UNEQUAL  |   INT32    |  10000000   |    100000    |   5.773 ms |       0.19% |   5.922 ms |       0.35% |   148.884 us |   2.58% |   SLOW   |
|     0      | NULLS_UNEQUAL  |   INT32    |  10000000   |   10000000   |  30.575 ms |       0.05% |  26.434 ms |       0.04% | -4141.727 us | -13.55% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |    1000     |     1000     | 143.676 us |       5.16% | 128.750 us |       4.01% |   -14.925 us | -10.39% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |   100000    |     1000     | 244.345 us |       5.78% | 193.447 us |       1.87% |   -50.899 us | -20.83% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |  10000000   |     1000     |  10.243 ms |       0.19% |   6.808 ms |       0.35% | -3435.389 us | -33.54% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |   100000    |    100000    | 282.226 us |       3.22% | 234.822 us |       1.68% |   -47.404 us | -16.80% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |  10000000   |    100000    |  13.462 ms |       0.15% |   9.045 ms |       0.06% | -4417.106 us | -32.81% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |  10000000   |   10000000   |  22.631 ms |       0.09% |  18.688 ms |       0.12% | -3943.154 us | -17.42% |   FAST   |

# mixed_left_join

## [0] Quadro RTX 8000

|  Nullable  |  NullEquality  |  DataType  |  left_size  |  right_size  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |         Diff |   %Diff |  Status  |
|------------|----------------|------------|-------------|--------------|------------|-------------|------------|-------------|--------------|---------|----------|
|     0      | NULLS_UNEQUAL  |   INT32    |    1000     |     1000     | 143.144 us |       5.89% | 127.381 us |       5.25% |   -15.763 us | -11.01% |   FAST   |
|     0      | NULLS_UNEQUAL  |   INT32    |   100000    |     1000     | 184.149 us |       2.65% | 166.885 us |       2.21% |   -17.264 us |  -9.37% |   FAST   |
|     0      | NULLS_UNEQUAL  |   INT32    |  10000000   |     1000     |   5.318 ms |       0.76% |   5.037 ms |       0.76% |  -280.916 us |  -5.28% |   FAST   |
|     0      | NULLS_UNEQUAL  |   INT32    |   100000    |    100000    | 206.159 us |       2.60% | 189.643 us |       4.08% |   -16.515 us |  -8.01% |   FAST   |
|     0      | NULLS_UNEQUAL  |   INT32    |  10000000   |    100000    |   6.288 ms |       0.47% |   5.651 ms |       0.41% |  -637.029 us | -10.13% |   FAST   |
|     0      | NULLS_UNEQUAL  |   INT32    |  10000000   |   10000000   |  29.465 ms |       0.11% |  27.118 ms |       0.03% | -2347.296 us |  -7.97% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |    1000     |     1000     | 145.859 us |       4.96% | 130.734 us |       2.34% |   -15.125 us | -10.37% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |   100000    |     1000     | 240.992 us |       4.67% | 200.980 us |       3.51% |   -40.012 us | -16.60% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |  10000000   |     1000     |   9.626 ms |       0.13% |   7.355 ms |       0.46% | -2270.471 us | -23.59% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |   100000    |    100000    | 274.480 us |       1.64% | 241.181 us |       1.52% |   -33.299 us | -12.13% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |  10000000   |    100000    |  12.782 ms |       0.10% |   9.510 ms |       0.06% | -3271.829 us | -25.60% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |  10000000   |   10000000   |  22.239 ms |       0.06% |  19.189 ms |       0.08% | -3049.949 us | -13.71% |   FAST   |

# mixed_full_join

## [0] Quadro RTX 8000

|  Nullable  |  NullEquality  |  DataType  |  left_size  |  right_size  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |         Diff |   %Diff |  Status  |
|------------|----------------|------------|-------------|--------------|------------|-------------|------------|-------------|--------------|---------|----------|
|     0      | NULLS_UNEQUAL  |   INT32    |    1000     |     1000     | 198.929 us |       2.34% | 185.816 us |       3.14% |   -13.113 us |  -6.59% |   FAST   |
|     0      | NULLS_UNEQUAL  |   INT32    |   100000    |     1000     | 217.827 us |       3.52% | 227.471 us |       2.05% |     9.644 us |   4.43% |   SLOW   |
|     0      | NULLS_UNEQUAL  |   INT32    |  10000000   |     1000     |   5.636 ms |       0.74% |   5.195 ms |       0.72% |  -440.313 us |  -7.81% |   FAST   |
|     0      | NULLS_UNEQUAL  |   INT32    |   100000    |    100000    | 273.253 us |       4.67% | 251.303 us |       1.63% |   -21.950 us |  -8.03% |   FAST   |
|     0      | NULLS_UNEQUAL  |   INT32    |  10000000   |    100000    |   6.509 ms |       0.48% |   5.795 ms |       0.42% |  -714.199 us | -10.97% |   FAST   |
|     0      | NULLS_UNEQUAL  |   INT32    |  10000000   |   10000000   |  31.266 ms |       0.07% |  28.127 ms |       0.05% | -3138.801 us | -10.04% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |    1000     |     1000     | 204.087 us |       3.91% | 189.799 us |       4.52% |   -14.287 us |  -7.00% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |   100000    |     1000     | 302.451 us |       3.88% | 259.927 us |       2.81% |   -42.525 us | -14.06% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |  10000000   |     1000     |  10.167 ms |       0.13% |   7.503 ms |       0.42% | -2664.698 us | -26.21% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |   100000    |    100000    | 339.111 us |       1.60% | 306.924 us |       3.05% |   -32.187 us |  -9.49% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |  10000000   |    100000    |  13.245 ms |       0.13% |   9.668 ms |       0.12% | -3577.098 us | -27.01% |   FAST   |
|     1      | NULLS_UNEQUAL  |   INT32    |  10000000   |   10000000   |  23.633 ms |       0.07% |  20.241 ms |       0.49% | -3391.283 us | -14.35% |   FAST   |
```

PointKernel commented:
/ok to test 75a018e

PointKernel commented on cpp/include/cudf/join/mixed_join.hpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2025, NVIDIA CORPORATION.
+ * Copyright (c) 2024-2025, NVIDIA CORPORATION.
```
```suggestion
 * Copyright (c) 2025, NVIDIA CORPORATION.
```

PointKernel commented on cpp/src/join/mixed_join_size_kernel.cuh:
```diff
@@ -29,119 +29,160 @@
-                                 std::size_t* output_size,
-                                 cudf::device_span<cudf::size_type> matches_per_row)
+#include <cuda/atomic>
+#include <thrust/reduce.h>
```
```suggestion
```

PointKernel commented on cpp/src/join/mixed_join_size_kernel.cuh:
```diff
@@ -29,119 +29,160 @@
-  // Allocate storage for the counter used to get the size of the join output
-  cudf::detail::device_scalar<std::size_t> size(0, stream, mr);
+  // Allocate device memory for the total count using the current device memory resource
+  rmm::device_scalar<size_t> d_total_count{0, stream, cudf::get_current_device_resource_ref()};
```
```suggestion
  cudf::detail::device_scalar<std::size_t> d_total_count{0, stream, cudf::get_current_device_resource_ref()};
```

PointKernel commented:
/ok to test c7b150c

GregoryKimball commented:
@pmattione-nvidia would you please share your review on the Java changes? Is Spark-RAPIDS going to need changes?

@wence- would you please help us with the core C++ changes and the python breaking change as well?

wence- commented on python/pylibcudf/pylibcudf/libcudf/join.pxd:
```diff
@@ -19,7 +19,7 @@ from pylibcudf.libcudf.utilities.span cimport device_span
 ctypedef unique_ptr[device_uvector[size_type]] gather_map_type
 ctypedef pair[gather_map_type, gather_map_type] gather_map_pair_type
-ctypedef optional[pair[size_t, device_span[const size_type]]] output_size_data_type
+ctypedef optional[size_t] output_size_data_type
```
Let's remove this typedef and use `optional[size_t]` directly in the forward declarations.

Can we also please rename the parameter from `output_size_data` to `output_size` (since it no longer takes the data as well).

So, for example:

```
cdef gather_map_pair_type mixed_left_join(
   ...
   optional[size_t] output_size,
   ...
)
```

wence- reviewed (changes requested):
Python part: requesting minor cleanup changes.

wence- commented on cpp/tests/join/mixed_join_tests.cu:
```diff
@@ -210,19 +210,12 @@ struct MixedJoinPairReturnTest : public MixedJoinTest<T> {
-                        nullptr,
-                        0);
-    CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected_counts_cw, actual_counts_view);
+    // Since we no longer get per-row counts, we can't verify them
+    // Just verify the total count matches
```
```suggestion
    // Since we no longer get per-row counts, we can't verify them.
    // Instead just verify that the total count matches
```

wence- commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -874,24 +874,16 @@ jlongArray mixed_join_size(JNIEnv* env,
+                                    condition->get_top_expression(),
+                                    nulls_equal);
+
+    // Return only the size value as a single element array
```
```suggestion
```

PointKernel commented on cpp/src/join/mixed_join.cu:
```diff
@@ -37,28 +37,96 @@
+  return std::make_pair(std::move(input_pairs), std::move(hash_indices));
+}
+
+}  // anonymous namespace
```
I find this particularly useful when a file contains multiple layers of nested namespaces, as it makes it easier to quickly identify the scope of a given function.

PointKernel commented on cpp/src/join/mixed_join_kernel.cuh:
```diff
@@ -29,37 +30,175 @@
+    size_type offset = 0;
+    auto const count = flushing_cg_counter[flushing_cg_id];
+    auto const rank  = tile.thread_rank();
+    if (rank == 0) { offset = atomic_counter.fetch_add(count, cuda::memory_order_relaxed); }
+    offset = tile.shfl(offset, 0);
```
This is essentially the same use case shown in the [CUDA user guide](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#broadcast-of-a-single-value-across-a-warp). And per [internal discussion](https://nvidia.slack.com/archives/C07DK328L/p1756317519875729), sync before shfl is not needed.

PointKernel commented:
@wence- Thanks for the thorough review. I believe I’ve addressed all your comments, though I may have overlooked one or two. Please let me know if I did.

wence- reviewed (approved):
Tiny issues, but I don't think I need to look over again. Approving C++ and Python.

PointKernel commented on cpp/src/join/hash_join.cu:
```diff
@@ -52,20 +52,7 @@ namespace {
- private:
-  Hasher _hash;
-};
+// TODO: `pair_equal` to be moved to common utils during mixed-join migration
```
Good catch. Many of the join utilities aren’t in their ideal files or locations. I initially planned to address all of them in this PR but realized it would affect numerous areas across the join implementations, so it’s better handled in a separate PR. I’ve updated the list in #19270 to ensure nothing gets overlooked.

PointKernel commented on cpp/src/join/hash_join.cu:
```diff
@@ -52,20 +52,7 @@ namespace {
- private:
-  Hasher _hash;
-};
+// TODO: `pair_equal` to be moved to common utils during mixed-join migration
```
https://github.com/rapidsai/cudf/pull/19660#discussion_r2317160680 it will be a task tracking separately.

ttnghia commented on java/src/main/java/ai/rapids/cudf/MixedJoinSize.java:
```diff
@@ -19,25 +19,22 @@
-    return matches;
-  }
-
+  /**
+   * Close this resource. This implementation does nothing as there are no resources to release,
+   * but implementing AutoCloseable allows this class to be used in try-with-resources blocks.
+   */
```
Since there is no resource to close, it should not be used in the `try` block. Any error/exception thrown out is still being handled properly. So we should not make this class `AutoClosable` anymore, to reduce mis-use.

ttnghia commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -874,24 +874,15 @@ jlongArray mixed_join_size(JNIEnv* env,
+                                    condition->get_top_expression(),
+                                    nulls_equal);
+
+    cudf::jni::native_jlongArray result(env, 1);
```
Yes, `MixedJoinSize.java` should be removed entirely and replaced by just a `long` number. I don't see other places that need to be changed.

PointKernel commented:
/merge
